ext/mysqli/mysqli_api.c

```diff
@@ -1695,31 +1705,30 @@ PHP_FUNCTION(mysqli_options)
 	}
 #endif
 	expected_type = mysqli_options_get_option_zval_type(mysql_option);
-	if (expected_type != Z_TYPE_P(mysql_value)) {
-		switch (expected_type) {
-			case IS_STRING:
-				if (!try_convert_to_string(mysql_value)) {
-					RETURN_THROWS();
-				}
-				break;
-			case IS_LONG:
-				convert_to_long_ex(mysql_value);
-				break;
-			default:
-				break;
+	if (expected_type == IS_STRING) {
+		if (!mysql_value_str) {
+			mysql_value_str = zend_long_to_str(mysql_value_long);
 		}
-	}
-	switch (expected_type) {
-		case IS_STRING:
-			ret = mysql_options(mysql->mysql, mysql_option, Z_STRVAL_P(mysql_value));
-			break;
-		case IS_LONG:
-			l_value = Z_LVAL_P(mysql_value);
-			ret = mysql_options(mysql->mysql, mysql_option, (char *)&l_value);
-			break;
-		default:
-			ret = 1;
-			break;
+		ret = mysql_options(mysql->mysql, mysql_option, ZSTR_VAL(mysql_value_str));
+
+		zend_string_release(mysql_value_str);
+	} else if (expected_type == IS_LONG) {
+		if (mysql_value_str) {
+			double rv;
+			zend_long lv;
+			zend_uchar type;
+
+			type = is_numeric_string(ZSTR_VAL(mysql_value_str), ZSTR_LEN(mysql_value_str), &lv, &rv, 0);
+			if (type == IS_LONG) {
+				mysql_value_long = lv;
+			} else {
+				zend_argument_type_error(getThis() ? 1 : 2, "must be a numeric string for the chosen option");
```

Comment:
This message is not correct, since it is thrown in case of decimal strings. And I'm also not sure about the TypeError.

Reply:
Hmm.. The i386 job using JIT fails with `gc_possible_root: Assertion `zval_gc_type((ref)->gc.u.type_info) == 7 || zval_gc_type((ref)->gc.u.type_info) == 8' failed` when running the `mysqli_option` test. What did I do wrong?

Reply:
Problem solved :) I had to conditionally free `mysql_value_str`.